{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from collections import defaultdict, Counter\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to build an index from the transcriptome data\n",
    "# Map k-mers to transcript IDs\n",
    "def build_index(transcriptome_file, k):\n",
    "    # Empty dictionary to store index (hash-table)\n",
    "    index = defaultdict(list) #Get list of transcripts that contain a given k-mer\n",
    "    \n",
    "    # get transcriptome \n",
    "    with open(transcriptome_file, \"r\") as handle:\n",
    "        # Parse each sequence in file\n",
    "        for record in SeqIO.parse(handle, \"fasta\"):\n",
    "            \n",
    "            # Convert sequence to string\n",
    "            seq = str(record.seq)\n",
    "            # Get transcript ID\n",
    "            transcript_id = record.id\n",
    "            \n",
    "            # Get all possible k-mers of length k\n",
    "            for i in range(len(seq) - k + 1):\n",
    "                # Extract k-mer at position i, with length k\n",
    "                kmer = seq[i:i+k]\n",
    "                \n",
    "                # Add transcript ID to hash-table for this k-mer\n",
    "                index[kmer].append(transcript_id) \n",
    "                    #ex ATC -> [isoform1, isoform2]\n",
    "    \n",
    "    return index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to compute the reverse complement of a sequence\n",
    "def reverse_complement(seq):\n",
    "    # Dictionary mapping each nucleotide to complement\n",
    "    complement = {\n",
    "        'A': 'T',  # Adenine -> Thymine\n",
    "        'C': 'G',  # Cytosine -> Guanine\n",
    "        'G': 'C',  # Guanine -> Cytosine\n",
    "        'T': 'A',  # Thymine -> Adenine\n",
    "        'N': 'N'   # Any base -> Any base (gets dealt with somewhere else)\n",
    "    }\n",
    "\n",
    "    # Empty string to store reverse complement sequence\n",
    "    reverse_comp_seq = ''\n",
    "\n",
    "    # Loop through sequence in reverse order\n",
    "    for base in reversed(seq):\n",
    "        # Add complement of current base to reverse complement sequence\n",
    "        reverse_comp_seq += complement[base]\n",
    "    \n",
    "    return reverse_comp_seq\n",
    "\n",
    "def generate_kmers(seq, k):\n",
    "    kmers = set()  # Set to store unique k-mers\n",
    "    \n",
    "    # Generate k-mers from the sequence\n",
    "    for i in range(len(seq) - k + 1):\n",
    "        kmer = seq[i:i + k]\n",
    "        if 'N' not in kmer:  # Skip k-mers with 'N'\n",
    "            kmers.add(kmer)\n",
    "    \n",
    "    return kmers\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pseudoalign(index, reads_file, k):\n",
    "    equivalence_classes = defaultdict(int)  # Dictionary to store equivalence classes and their counts\n",
    "    \n",
    "    with open(reads_file, \"rt\") as handle:  # Open the reads file\n",
    "        for record in SeqIO.parse(handle, \"fasta\"):  # For each sequence in the reads file\n",
    "            read_seq = str(record.seq)  # Convert sequence to string\n",
    "            rc_read_seq = reverse_complement(read_seq)  # Reverse complement of the sequence\n",
    "            \n",
    "            matched_transcripts = None  # Initialize matched transcripts as None\n",
    "            \n",
    "            # Generate k-mers from read (forward strand)\n",
    "            kmers_read = generate_kmers(read_seq, k)\n",
    "            # Generate k-mers from reverse complement of the read\n",
    "            kmers_rc_read = generate_kmers(rc_read_seq, k)\n",
    "            \n",
    "            # Get matches from forward sequence\n",
    "            for kmer in kmers_read:\n",
    "                if kmer in index:\n",
    "                    if matched_transcripts is None: # If no matches\n",
    "                        matched_transcripts = set(index[kmer]) # Initialize set of matched transcripts\n",
    "                    else: # If there are matches\n",
    "                        matched_transcripts.intersection_update(index[kmer]) # Update set of matched transcripts\n",
    "            \n",
    "            # Get matches from reverse complement sequence\n",
    "            for kmer in kmers_rc_read:\n",
    "                if kmer in index:\n",
    "                    if matched_transcripts is None:\n",
    "                        matched_transcripts = set(index[kmer])\n",
    "                    else:\n",
    "                        matched_transcripts.intersection_update(index[kmer])\n",
    "            \n",
    "            #print(f\"Matched transcripts: {matched_transcripts}\")\n",
    "            \n",
    "            if matched_transcripts and len(matched_transcripts) > 0:  # If there are matches, create an equivalence class\n",
    "                eq_class = tuple(sorted(matched_transcripts))\n",
    "            else:  # If no matches, then 'NA' as equivalence class\n",
    "                eq_class = ('NA',)\n",
    "            \n",
    "            # Update count of THIS equivalence class\n",
    "            equivalence_classes[eq_class] += 1\n",
    "    \n",
    "    return equivalence_classes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Main function to return results\n",
    "def run_pseudoalignment(transcriptome_file, reads_file, k):\n",
    "    index = build_index(transcriptome_file, k) #Get index made of k-mers length k\n",
    "    pseudoalignment_results = pseudoalign(index, reads_file, k) #Pseudoalign\n",
    "    \n",
    "    # Output results\n",
    "    for eq_class, count in pseudoalignment_results.items():\n",
    "        num_items = len(eq_class) # Number of isoforms in equivalence class\n",
    "        isoforms = \",\".join(eq_class) #Join isoforms into single string\n",
    "        if not isoforms: #If no isoforms, set to NA\n",
    "            isoforms = 'NA'\n",
    "        # print(f\"{count}\\t{num_items}\\t{isoforms}\") #Print count, # items, and isoforms\n",
    "    \n",
    "    # statistics\n",
    "    eq_class_sizes = {} #Empty dictionary to count sizes of equivalence classes\n",
    "\n",
    "    # Loop through each equivalence class in pseudoalignment results\n",
    "    for eq_class in pseudoalignment_results:\n",
    "        # size of current equivalence class\n",
    "        size = len(eq_class)\n",
    "        \n",
    "        # If size already in dictionary, increment its count\n",
    "        if size in eq_class_sizes:\n",
    "            eq_class_sizes[size] += 1\n",
    "        # If size is not in dictionary, add it with a count of 1\n",
    "        else:\n",
    "            eq_class_sizes[size] = 1\n",
    "\n",
    "    # Loop through each size and its count in eq_class_sizes dictionary\n",
    "    # for size, count in eq_class_sizes.items():\n",
    "    #     # Print size of equivalence class and number of classes with that size\n",
    "    #     print(\"Equivalence class size\", size, \":\", count, \"classes\")\n",
    "        \n",
    "    return pseudoalignment_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def equivalence_classes_to_dataframe(equivalence_classes):\n",
    "    data = {\n",
    "        \"counts\": [],\n",
    "        \"number of items in equivalence class\": [],\n",
    "        \"isoforms in equivalence class\": []\n",
    "    } #Empty dictionary to store data\n",
    "    \n",
    "    for eq_class, count in equivalence_classes.items(): #Loop through each equivalence class\n",
    "        num_items = len(eq_class)\n",
    "        isoforms = \",\".join(eq_class) if eq_class else \"NA\" #Join isoforms into single string\n",
    "        \n",
    "        data[\"counts\"].append(count) #Append count\n",
    "        data[\"number of items in equivalence class\"].append(num_items) #Append number of items\n",
    "        data[\"isoforms in equivalence class\"].append(isoforms) #Append isoforms \n",
    "    \n",
    "    df = pd.DataFrame(data)\n",
    "    df = df.sort_values(by=\"counts\", ascending=False) #Sort by descending order\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameters\n",
    "transcriptome_file = \"/Users/timothyliu/Documents/121/Project/chr11_transcriptome.fasta\"\n",
    "reads_file = \"/Users/timothyliu/Documents/121/Project/reads.fasta\"\n",
    "sam_file = \"aligned_reads.sam\"\n",
    "k = 31  # k-mer length\n",
    "\n",
    "# Run pseudoalignment implementation\n",
    "pseudoalignment_results = run_pseudoalignment(transcriptome_file, reads_file, k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "      <th>number of items in equivalence class</th>\n",
       "      <th>isoforms in equivalence class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>66897</td>\n",
       "      <td>1</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4302</th>\n",
       "      <td>61263</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000329251,ENST00000496634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1737</th>\n",
       "      <td>15353</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000536684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2162</th>\n",
       "      <td>12320</td>\n",
       "      <td>7</td>\n",
       "      <td>ENST00000227157,ENST00000379412,ENST0000039622...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8907</th>\n",
       "      <td>12097</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000393067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3871</th>\n",
       "      <td>10440</td>\n",
       "      <td>5</td>\n",
       "      <td>ENST00000345732,ENST00000389939,ENST0000053207...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>547</th>\n",
       "      <td>10038</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000321153,ENST00000530398,ENST00000530797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1989</th>\n",
       "      <td>9317</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000228140,ENST00000525634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9590</th>\n",
       "      <td>8807</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000527673,ENST00000532567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9885</th>\n",
       "      <td>7457</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000227378,ENST00000524552,ENST0000052611...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9861</th>\n",
       "      <td>7386</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000260197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9589</th>\n",
       "      <td>6755</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000527673,ENST00000527791,ENST00000532567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2161</th>\n",
       "      <td>6028</td>\n",
       "      <td>11</td>\n",
       "      <td>ENST00000227157,ENST00000379412,ENST0000039622...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9594</th>\n",
       "      <td>5954</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000527673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1465</th>\n",
       "      <td>5545</td>\n",
       "      <td>7</td>\n",
       "      <td>ENST00000314138,ENST00000524496,ENST0000052656...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      counts  number of items in equivalence class  \\\n",
       "32     66897                                     1   \n",
       "4302   61263                                     2   \n",
       "1737   15353                                     1   \n",
       "2162   12320                                     7   \n",
       "8907   12097                                     1   \n",
       "3871   10440                                     5   \n",
       "547    10038                                     3   \n",
       "1989    9317                                     2   \n",
       "9590    8807                                     2   \n",
       "9885    7457                                     8   \n",
       "9861    7386                                     1   \n",
       "9589    6755                                     3   \n",
       "2161    6028                                    11   \n",
       "9594    5954                                     1   \n",
       "1465    5545                                     7   \n",
       "\n",
       "                          isoforms in equivalence class  \n",
       "32                                                   NA  \n",
       "4302                    ENST00000329251,ENST00000496634  \n",
       "1737                                    ENST00000536684  \n",
       "2162  ENST00000227157,ENST00000379412,ENST0000039622...  \n",
       "8907                                    ENST00000393067  \n",
       "3871  ENST00000345732,ENST00000389939,ENST0000053207...  \n",
       "547     ENST00000321153,ENST00000530398,ENST00000530797  \n",
       "1989                    ENST00000228140,ENST00000525634  \n",
       "9590                    ENST00000527673,ENST00000532567  \n",
       "9885  ENST00000227378,ENST00000524552,ENST0000052611...  \n",
       "9861                                    ENST00000260197  \n",
       "9589    ENST00000527673,ENST00000527791,ENST00000532567  \n",
       "2161  ENST00000227157,ENST00000379412,ENST0000039622...  \n",
       "9594                                    ENST00000527673  \n",
       "1465  ENST00000314138,ENST00000524496,ENST0000052656...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = equivalence_classes_to_dataframe(pseudoalignment_results)\n",
    "df.to_csv('/Users/timothyliu/Documents/121/Project/output.csv', index=False)\n",
    "df.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>unique</th>\n",
       "      <th>top</th>\n",
       "      <th>freq</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>counts</th>\n",
       "      <td>10489.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>122.273429</td>\n",
       "      <td>1007.425181</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>66897.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>number of items in equivalence class</th>\n",
       "      <td>10489.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.079321</td>\n",
       "      <td>5.316185</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>54.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoforms in equivalence class</th>\n",
       "      <td>10489</td>\n",
       "      <td>10489</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        count unique  top freq        mean  \\\n",
       "counts                                10489.0    NaN  NaN  NaN  122.273429   \n",
       "number of items in equivalence class  10489.0    NaN  NaN  NaN    5.079321   \n",
       "isoforms in equivalence class           10489  10489   NA    1         NaN   \n",
       "\n",
       "                                              std  min  25%   50%   75%  \\\n",
       "counts                                1007.425181  1.0  2.0  10.0  46.0   \n",
       "number of items in equivalence class     5.316185  1.0  2.0   4.0   7.0   \n",
       "isoforms in equivalence class                 NaN  NaN  NaN   NaN   NaN   \n",
       "\n",
       "                                          max  \n",
       "counts                                66897.0  \n",
       "number of items in equivalence class     54.0  \n",
       "isoforms in equivalence class             NaN  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe(include='all').T"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py311",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
